File: pset6/credit.py
from cs50 import get_int

# prompt for card's cc_numbers
while True:
    credit_card = get_int("Credit Card Number: ")
    if credit_card > 0:
        break

# convert to string and get the card_size of the cc_numbers string
card_size = len(str(credit_card))

cc_numbers = list()
divisor = 1

# store each number in array
for i in str(credit_card):
    cc_numbers.append(int(i))

# reverse order
cc_numbers = cc_numbers[::-1]

# 1st sum
sum = 0
for i in range(1, card_size, 2):
    # 2 cases: product < 10 and >=10
    if cc_numbers[i] * 2 < 10:
        sum = sum + cc_numbers[i] * 2
    else:
        sum = sum + (cc_numbers[i] * 2) % 10 + ((cc_numbers[i] * 2) // 10) % 10

# 2nd sum
for i in range(0, card_size, 2):
    sum += cc_numbers[i]

# check the validity of the card number
if (sum % 10) != 0:
    print("INVALID")
else:
    if card_size == 15 and cc_numbers[card_size - 1] == 3 and (cc_numbers[card_size - 2] == 4 or cc_numbers[card_size - 2] == 7):
        print("AMEX")
    elif card_size == 16 and cc_numbers[card_size - 1] == 5 and (cc_numbers[card_size - 2] == 1 or cc_numbers[card_size - 2] == 2 or cc_numbers[card_size - 2] == 3 or cc_numbers[card_size - 2] == 4 or cc_numbers[card_size - 2] == 5):
        print("MASTERCARD")
    elif (card_size == 13 or card_size == 16) and cc_numbers[card_size - 1] == 4:
        print("VISA")
    else:
        print("INVALID")

File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    long card_num = 0LL;

    while (card_num < 1LL || card_num > 9999999999999999LL)
    {
        card_num = get_long("Please give me your credit card number: ");
    }

    // Make a copy of the card number to be used and modified throughout the process.

    long card_num_copy = card_num;

    // Isolate every digit from the credit card number using a loop and the variable 'digit'.
    // Keep track of the amount and position of each digit using variable 'count'.

    int card_num_count = 0;

    while (card_num_copy > 0LL)
    {
        card_num_copy = card_num_copy / 10LL;
        card_num_count++;
    }

    // If the number entered doesn't have the right amount of digits according
    // to variable 'count', declare the number as invalid.

    if (card_num_count != 13 && card_num_count != 15 && card_num_count != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    // Reset value of variable 'card_num_copy' and apply calculations that will isolate the first two digits.
    // Store the results in a variable 'company_id'.

    card_num_copy = card_num;

    while (card_num_copy > 100LL)
    {
        card_num_copy = card_num_copy / 10LL;
    }

    int company_id = card_num_copy;

    // Print the type of credit card depending on the company ID and amount of digits.

    if (company_id > 50 && company_id < 56 && card_num_count == 16)
    {
        printf("MASTERCARD\n") ;
    }
    else if ((company_id == 34 || company_id == 37) && (card_num_count == 15))
    {
        printf("AMEX\n") ;
    }
    else if ((company_id / 10 == 4) && (card_num_count == 13 || card_num_count == 16))
    {
        printf("VISA\n") ;
    }
    else
    {
        printf("INVALID\n");
    }

    // Luhn's algorithm.

    int sum = 0;

    card_num_copy = card_num;

    for (int i = 1; i <= card_num_count; i++)
    {
        int digit = card_num_copy % 10LL;

        if (i % 2 == 0)
        {
            digit *= 2;

            if (digit > 9)
            {
                digit -= 9;
            }
        }

        sum += digit;

        card_num_copy /= 10LL;
    }

    // Checking the validity of the number according to Luhn's algorithm

    if (sum % 10 != 0)
    {
        printf("INVALID\n");
    }
    return 0;
}
